Nav fix (#1542)
SumoLogic/sumologic-documentation: docs/integrations/global-intelligence/kubernetes-devops.md

**Collection Process Overview:** With this graphic, you can see how to collect metrics from your Kubernetes environment and send them to Sumo Logic.

![img](/img/kubernetes/collection-process-overview.png)

Follow the steps on the [Sumo Logic Kubernetes Deployment guide](https://github.com/SumoLogic/sumologic-kubernetes-collection/tree/main/deploy#metrics) to configure the collection for Kubernetes Benchmark metrics.

## Installing the Global Intelligence for Kubernetes DevOps App and Viewing Dashboards

Below are the instructions for installing the Sumo App for Kubernetes DevOps as well as descriptions of each of the app dashboards. 

### Dashboard Filters  

**Each dashboard has a set of filters** that you can apply to the entire dashboard, as shown in the following example. Click the funnel icon in the top dashboard menu bar to display a scrollable list of filters that are applied across the entire dashboard. 

:::note
You can use filters to drill down and examine the data on a granular level. 
:::

![img](/img/kubernetes/kubernetes-dashboard-filter.png)

> Select a **cluster**, **namespace**, and **deployment**. The **history_percent** represents the percent of 15m epocks over the prior 8 days over which Kubernetes metrics data was available for analysis for a given deployment-container combination. Percentages less than 100% represent deployment-container combinations that were new, transient or otherwise inactive for some portion of the prior 8 days.

Reviewer: @kimsauce did you mean "epochs"?

Author: Oh I didn't actually write this 😄! I just moved it from one doc to another, but didn't notice that. Yikes and thank you!

> Select a **cluster**, **namespace**, and **deployment**. The **history_percent** represents the percent of 15m epocks over the prior 8 days over which Kubernetes metrics data was available for analysis for a given deployment-container combination. Percentages less than 100% represent deployment-container combinations that were new, transient or otherwise inactive for some portion of the prior 8 days.

Author: ```suggestion
Select a **cluster**, **namespace**, and **deployment**. The **history_percent** represents the percent of 15m epochs over the prior 8 days over which Kubernetes metrics data was available for analysis for a given deployment-container combination. Percentages less than 100% represent deployment-container combinations that were new, transient or otherwise inactive for some portion of the prior 8 days.
```

> Select a **cluster**, **namespace**, and **deployment**. The **history_percent** represents the percent of 15m epocks over the prior 8 days over which Kubernetes metrics data was available for analysis for a given deployment-container combination. Percentages less than 100% represent deployment-container combinations that were new, transient or otherwise inactive for some portion of the prior 8 days.

Reviewer: ```suggestion
Select a **cluster**, **namespace**, and **deployment**. The **history_percent** represents the percent of 15m epochs over the prior 8 days over which Kubernetes metrics data was available for analysis for a given deployment-container combination. Percentages less than 100% represent deployment-container combinations that were new, transient or otherwise inactive for some portion of the prior 8 days.
```

The `infer` operator is the interface to a machine learning model. It is intended for use only in this application.

The infrastructure tab of the [Entity Inspector](../../dashboards-new/drill-down-to-discover-root-causes.md) is not compatible with Global Intelligence for Kubernetes DevOps and renders a blank page on Dashboards Summary for Kubernetes Deployment and Detail for Kubernetes Deployment. This is because Global Intelligence apps analyze anonymized customer data and do not have a way to identify customer entities required for Entity Inspector. 

Deployment-container combinations that have no requests or limits set are not candidates for optimization. No recommendations are computed in these cases. 
:::

### Concepts, Definitions and Methodology

It is relatively well known that DevOps staff concerned about Out of Memory (OOM) and CPU throttling may set high memory and CPU requests on their containers. This results in situations where Kubernetes clusters are over-provisioned compared to actual usage. Conversely, Kubernetes clusters may be underprovisioned, leading to risk of OOM and throttling and resulting downtime. 

> The slack is the difference between usage and requests measures cost for both cpu (cores) and memory (GB). Higher slack implies higher costs as cpu and memory used are less than what has been provisioned for a given container. Conversely negative slack, while saving costs, increase risk for container eviction due to memory overages or container throttling. These latter events impact application availability and performance and may lead to incidents. The cost-risk tradeoff is diagrammed below. 

Reviewer: ```suggestion
The slack is the difference between usage and requests measures cost for both cpu (cores) and memory (GB). Higher slack implies higher costs as cpu and memory used are less than what has been provisioned for a given container. Conversely, negative slack, while saving costs, increased the risk of container eviction due to memory overages or container throttling. These latter events impact application availability and performance and may lead to incidents. The cost-risk tradeoff is diagrammed below. 
```

Reviewer: @kimsauce The first sentence (starting with "The slack is") doesn't make sense to me, in particular "requests measures cost". Maybe there's an extra or missing word?

The input for the optimization is time series data for each container’s cpu, memory usage, and requests over the past eight days. A given container can behave differently depending on the pod in which it is deployed. For example, an nginx container in Pod A may behave differently from the same container in Pod B. As a result, GI Kubernetes DevOps detects unique deployment-container combinations called "container groups", analyzes cpu and memory time series data for them and recommends optimal requests and limits based on the best combination of slack (or cost) and risk. The recommendations cover three scenarios; pick the one that best serves the needs of your application:

> * **Cost-optimized**: lowest slack (or cost) accompanied with relatively higher risk (measured by percent at-risk minutes for the deployment-container)

Reviewer: ```suggestion
* **Cost-optimized**.  Lowest slack (or cost) accompanied with relatively higher risk (measured by percent at-risk minutes for the deployment-container)
```

> * **Cost-optimized**.  Lowest slack (or cost) accompanied with relatively higher risk (measured by percent at-risk minutes for the deployment-container)

Author: ```suggestion
* **Cost-optimized**. Lowest slack (or cost) accompanied with relatively higher risk (measured by percent at-risk minutes for the deployment-container)
```

> * **Reliability-optimized**: highest slack (or cost) accompanied with relatively lower risk (measured by percent at-risk minutes for the deployment-container)

Reviewer: ```suggestion
* **Reliability-optimized**. Jighest slack (or cost) accompanied with relatively lower risk (measured by percent at-risk minutes for the deployment-container)
```

> * **Balanced**: strikes the balance between cost and risk; cost is relatively higher than the cost-optimized scenario while risks are relatively lower.

Reviewer: ```suggestion
* **Balanced**. Strikes the balance between cost and risk; cost is relatively higher than the cost-optimized scenario while risks are relatively lower.
```

![costs](/img/kubernetes/kubernetes-cost.png)

Time series data for usage and requests is summarized for each deployment-container combination to compute slack as follows:

* Time series are bucketed into 15m epochs and summarized using 15m averages.

> * Usage is represented by the p95 value over the calendar day over all deployment-container combinations. Let us call this `usage_15m_p95_1d(deployment_container)`.

Reviewer: ```suggestion
* Usage is represented by the p95 value over the calendar day over all deployment-container combinations. Let's call this `usage_15m_p95_1d(deployment_container)`.
```

> * Request is represented by the average over the calendar day of 15m epochs for each deployment-container combination. Let us call this `request_15m_avg_1d(deployment_container)`.

Reviewer: ```suggestion
* Request is represented by the average over the calendar day of 15m epochs for each deployment-container combination. Let's call this `request_15m_avg_1d(deployment_container)`.
```

* Next, slack for a calendar day in any 15m epoch for any deployment-container combination is calculated as `slack_15m_1d(deployment_container) =  request_15m_avg_1d(deployment_container) -  usage_15m_p95_1d(deployment_container)`

> * Then, we calculate `slack_8d(deployment_container)` as average over  `slack_15m_1d(deployment_container)`. `slack_8d(deployment_container)` is an estimate of the slack in any 15m interval for a given deployment-container combination based on 8d data.

Reviewer: ```suggestion
* We calculate `slack_8d(deployment_container)` as average over `slack_15m_1d(deployment_container)`. `slack_8d(deployment_container)` is an estimate of the slack in any 15m interval for a given deployment-container combination, based on 8d data.
```
